contrib/pkg/v1migration/recreateobjects.go

```diff
@@ -0,0 +1,159 @@
+package v1migration
+
+import (
+	"bufio"
+	"encoding/json"
+	"io"
+	"os"
+
+	"github.com/pkg/errors"
+	log "github.com/sirupsen/logrus"
+	"github.com/spf13/cobra"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/client-go/dynamic"
+
+	contributils "github.com/openshift/hive/contrib/pkg/utils"
+	hivev1alpha1 "github.com/openshift/hive/pkg/apis/hive/v1alpha1"
+)
+
+// RecreateObjectsOptions is the set of options for the re-creating Hive objects.
+type RecreateObjectsOptions struct {
+	fileName string
+}
+
+// NewRecreateObjectsCommand creates a command that executes the migration utility to re-create Hive objects stored in a file.
+func NewRecreateObjectsCommand() *cobra.Command {
+	opt := &RecreateObjectsOptions{}
+	cmd := &cobra.Command{
+		Use:   "recreate-objects JSON_FILE_NAME",
+		Short: "re-create the Hive objects stored in the JSON file",
+		Args:  cobra.ExactArgs(1),
+		Run: func(cmd *cobra.Command, args []string) {
+			log.SetLevel(log.InfoLevel)
+			if err := opt.Complete(cmd, args); err != nil {
+				log.WithError(err).Fatal("Error")
+			}
+
+			if err := opt.Validate(cmd); err != nil {
+				log.WithError(err).Fatal("Error")
+			}
+
+			if err := opt.Run(); err != nil {
+				log.WithError(err).Fatal("Error")
+			}
+		},
+	}
+	return cmd
+}
+
+// Complete finishes parsing arguments for the command
+func (o *RecreateObjectsOptions) Complete(cmd *cobra.Command, args []string) error {
+	o.fileName = args[0]
+	return nil
+}
+
+// Validate ensures that option values make sense
+func (o *RecreateObjectsOptions) Validate(cmd *cobra.Command) error {
+	return nil
+}
+
+// Run executes the command
+func (o *RecreateObjectsOptions) Run() error {
+	clientConfig, err := contributils.GetClientConfig()
+	if err != nil {
+		return errors.Wrap(err, "could not get the client config")
+	}
+	client, err := dynamic.NewForConfig(clientConfig)
+	if err != nil {
+		return errors.Wrap(err, "could not create kube client")
+	}
+	file, err := os.Open(o.fileName)
+	if err != nil {
+		errors.Wrap(err, "could not open the JSON file")
+	}
+	defer file.Close()
+	decoder := json.NewDecoder(bufio.NewReader(file))
+	for {
+		var objFromFile unstructured.Unstructured
+		if err := decoder.Decode(&objFromFile); err != nil {
+			if err == io.EOF {
+				break
+			}
+			return errors.Wrap(err, "could not decode JSON from file")
+		}
+		o.recreateObject(client, &objFromFile)
```

Comment:
Should we re-order HiveConfig specifically to be first here to avoid the problem I hit in migration when using managed DNS where hiveadmission will reject clusters because their domain isn't in the config yet? Or is this handled elsewhere? I will do another migration runthrough soon.

Reply:
Making it first won't help, unfortunately. We need the hive-operator to be running in order for the HiveConfig change to take effect. So we need to restore the HiveConfig as a separate action prior to scaling down. I will add this to my issue list.

Reply:
Ok feel free to dump any issues to jira tasks and we can farm them out as well if you'd backup.